deepmd/pt/utils/tabulate.py

```
@@ -0,0 +1,920 @@
+        for item in self.matrix["layer_" + str(self.layer_size)]:
+            if len(item) != 0:
+                return item.shape[1]
+        return 0
+    
+    def _convert_numpy_to_tensor(self):
+        """Convert self.data from np.ndarray to torch.Tensor."""
+        for ii in self.data:
+            self.data[ii] = torch.tensor(self.data[ii])
+
+    @property
+    @lru_cache
```

Ruff

Use of `functools.lru_cache` or `functools.cache` on methods can lead to memory leaks

(B019)